Changing skb_load_bytes to implement direct packet access.

trvll opened the pull request:
- Data needs to be verified with revalidate_data().
- Implements direct packet access when loading bytes from ctx.

Signed-off-by: Thales Paiva <thales@accuknox.com>


Fixes: #10808

borkmann commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -33,28 +33,23 @@ static __always_inline __maybe_unused int
-		     : [ret]"=r"(ret), [from]"=r"(from)
-		     : [ctx]"r"(ctx), [off]"r"(off), [len]"ri"(len),
-		       [offmax]"i"(__CTX_OFF_MAX), [errno]"i"(-EINVAL)
-		     : "r1", "r2");
```
Removing the inline asm here should likely result in verifier issues due to the way LLVM is generating the code. Originially, I had it similarly as you below, but had to switch to inline asm to force the pattern matching.

trvll replied:
@borkmann You are right, Thank you for this advice. I'm testing a new fix and will update that.

pchaigno reviewed (changes requested):
`xdp_load_bytes` already implements direct packets access. The required changes are on `skb_load_bytes`.

trvll committed:
Implementing direct packet load for skb_load_bytes.

As in XDP, skb_load_bytes() must implement direct packet access.
As ctx_load_bytes() translates to xdp_load_bytes() or skb_load_bytes()
depending on the context this will make it use direct packet access
in both contexts.

Signed-off-by: Thales Paiva <thales@accuknox.com>

borkmann commented on bpf/include/bpf/ctx/skb.h:
```diff
@@ -56,6 +57,37 @@
+		memcpy(to, from, len);
+	return ret;
+}
+#endif
```
Unfortunately this is not a direct equivalent of `bpf_skb_load_bytes()`. The reason is that the latter uses `skb_header_pointer()` which can query the skb linear data or non-linear parts. The latter is done via `skb_copy_bits ()` without pulling data into linear section.

 - https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/core/filter.c#n1720
 - https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n3681

The implementation above is only looking at the linear data, but not non-linear one. In case of XDP, there is only linear data (but not non-linear one, hence the asm inline equivalent in case of XDP).

The case of https://github.com/cilium/cilium/issues/10808 is saying that instead of implementing `skb_load_bytes()` via inline asm where internally does the `memcpy()` in your case, we should rather successively replace individual calls to `skb_load_bytes()`(here: meaning helper call) with direct `ctx->data` access like elsewhere in our code.

trvll replied:
Hi @borkmann Thank you so much for making it clearer I was in a wrong path since the beginning. I still have some doubt. Replacing individual calls by direct access (ctx->data and ctx->data_end) I would use a direct assignment with a correspondent casting for destination structure instead of using memcpy. Doing this way, the non-linear data parts will be covered anyway?

borkmann replied:
See for example (just to mention one sample) how https://github.com/cilium/cilium/blob/master/bpf/lib/nodeport.h#L201 data/data_end & `revalidate_data()` is used to assign the ip6 pointer and then read from it.

trvll replied:
Thanks for sharing. I'm working on it.
